### PR TITLE
txmetrics: Add `10m`, `1h` aggregation intervals

### PR DESCRIPTION
**Pull request by marclop**
## Motivation/summary

Modifies the existing `txmetrics` aggregator to support a list of extra
aggregation intervals by introducing a new `RollUpIntervals` (internal
only) config option. The configuration option is not user facing, but
rather allow us to configure smaller intervals for testing purposes.
Also removes the `aggregation.transactions.interval` config option.

Two new intervals are added and published by default, 10 and 60m (1h).
and these are flushed when the APM Server is stopped or restarted.

Introduces a new `metrics-apm.transaction.${interval}m` datastream
which contains the mapping definitions for transaction metrics with a
series of changes to the apmpackage generator to allow interval
templating and individual ILM policies to be tied to each interval.
A new `Metricset.Interval` field is added to the ingested documents, to
allow them to contain the aggregated interval they represent. The field
format is `${interval}m` or `${interval}s` if < 60s.

Older APM Servers sending data to an Elasticsearch cluster with updated
APM Package, will continue indexing metrics to `metrics-apm.internal`.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [x] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [x] Documentation has been updated

## How to test these changes

1. `tilt up`
2. Run the `test.sh` script in https://gist.github.com/marclop/01a8f4270cc442573fd8319ca140015d.
3. Toggle expvar in the APM integration so metrics are published immediately on hot restart.
4. Verify that 3 metric sets per application and data stream `metrics-apm.transaction.1m`, `metrics-apm.transaction.10m`, `metrics-apm.transaction.60m`, and that they contain `event.duration` in nanoseconds matching the data stream interval (`constant_keyword`).
```
GET /metrics-apm*/_search
{
  "_source": [false],
  "fields": [
    "metricset.name",
    "event.duration",
    "service.name",
    "metricset.interval"
  ]
}
```
5. Verify that each of the different data streams has a dedicated ILM policy.

## Related issues

Part of #9703
Depends on https://github.com/elastic/apm-data/pull/7

### CODE REVIEW

File: apmpackage/apm/data_stream/internal_metrics_interval/manifest.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+title: APM internal metrics
+type: metrics
+dataset: apm.internal-interval
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics_interval/manifest.yml:3]: [`data_stream.dataset`](https://www.elastic.co/guide/en/ecs/current/ecs-data_stream.html#field-data-stream-dataset) is not allowed to contain "-", as that's the separator used for identifying the data stream type, dataset, and namespace.

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go [hunk 2]
```diff
@@ -55,7 +55,9 @@ type Aggregator struct {
 	tooManyGroupsLogger *logp.Logger
 
 	mu               sync.RWMutex
-	active, inactive *metrics
+	active, inactive map[time.Duration]*metrics
```
Comment (axw) [inline x-pack/apm-server/aggregation/txmetrics/aggregator.go:58]: Given that we iterate over the intervals in order, would it be worth making these `[]*metrics`?

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go [hunk 3]
```diff
@@ -141,16 +160,28 @@ func (a *Aggregator) Run() error {
...
-				"publishing transaction metrics failed: %s", err,
-			)
+		// Publish the metricsets for all configured intervals.
+		for _, interval := range a.intervals {
+			// Publish $interval MetricSets when:
+			//  - ticks * MetricsInterval % $interval == 0.
+			//  - Aggregator is stopped.
+			if !stop && (ticks*uint64(a.config.MetricsInterval))%uint64(interval) != 0 {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (marclop) [reply x-pack/apm-server/aggregation/txmetrics/aggregator.go:178]: I did have that check initially and removed it since we aren't planning  to allow users to tweak that setting it seemed superfluous, I'm happy adding it back.

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go (hunk 2 above)
Comment (marclop) [reply x-pack/apm-server/aggregation/txmetrics/aggregator.go:58]: I suppose we could, would you prefer that?

Comment (axw) [reply x-pack/apm-server/aggregation/txmetrics/aggregator.go:58]: I don't have a very strong opinion on this - slight preference because it'll be slightly faster. Not sure if it's worth it. If it makes things harder to read then don't worry about it.

File: apmpackage/apm/data_stream/internal_metrics_interval/manifest.yml (hunk 1 above)
Comment (felixbarny) [reply apmpackage/apm/data_stream/internal_metrics_interval/manifest.yml:3]: > I would strongly prefer if we introduced transaction and service metric data streams.

I’m good with separating out data streams for metrics that we’re downsampling - transaction, service, and service destination metrics.

Comment (marclop) [reply apmpackage/apm/data_stream/internal_metrics_interval/manifest.yml:3]: So we'd have 1 internal metric data stream (`apm-metrics.internal`), and 3 "aggregation" metrics data streams (1 per aggregated interval) then.

File: apmpackage/apm/data_stream/internal_metrics_transaction/fields/faas_metrics.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+- name: faas.duration
+  type: float
+  metric_type: counter
+  unit: ms
+  description: The total number of milliseconds elapsed during the execution of the FaaS invocation.
+- name: faas.timeout
+  type: long
+  metric_type: counter
+  unit: ms
+  description: The FaaS invocation timeout, expressed in milliseconds.
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics_transaction/fields/faas_metrics.yml:20]: Are these relevant to transaction metrics? Doesn't look like it, looking at txmetrics code.

File: apmpackage/apm/data_stream/internal_metrics_transaction/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  description: |
+    Request ID of the function invocation.
+- name: faas.trigger.request_id
+  type: keyword
+  description: |
+    The ID of the origin trigger request.
+- name: faas.trigger.type
+  type: keyword
+  description: |
+    The trigger type.
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics_transaction/fields/fields.yml:20]: I don't think these are relevant either

File: apmpackage/apm/data_stream/internal_metrics_transaction/fields/fields.yml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+- name: faas.version
+  type: keyword
+  description: |
+    The lambda function version.
+- name: kubernetes.pod.name
+  type: keyword
+  description: |
+    Kubernetes pod name
+- name: metricset.name
+  type: keyword
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics_transaction/fields/fields.yml:34]: Can we make this `constant_keyword`, and set the value to "transaction" here?

**Review by axw — COMMENTED** [review]
Looking good! As discussed on Slack, we'll need to handle transaction metrics from older servers going to `metrics-apm.internal-*`, and either upgrade and re-route, or drop them.

File: apmpackage/apm/data_stream/internal_metrics_transaction/fields/faas_metrics.yml (hunk 4 above)
Comment (marclop) [reply apmpackage/apm/data_stream/internal_metrics_transaction/fields/faas_metrics.yml:20]: absolutely right, I'll delete these.

File: apmpackage/apm/data_stream/internal_metrics_transaction/fields/fields.yml (hunk 5 above)
Comment (marclop) [reply apmpackage/apm/data_stream/internal_metrics_transaction/fields/fields.yml:20]: `faas.trigger.type` is used, but the other two aren't, I'll remove them.

File: apmpackage/apm/data_stream/internal_metrics/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -1,6 +1,8 @@
 ---
 description: Pipeline for ingesting APM internal metrics.
 processors:
+  - pipeline: # drop transaction metrics coming from older APM Server versions.
+      name: drop_tx_metrics
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics/elasticsearch/ingest_pipeline/default.yml:5]: Should we inline the pipeline here? It's not used anywhere else.

File: apmpackage/apm/data_stream/internal_metrics_transaction/manifest.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+title: APM transaction metrics
+type: metrics
+dataset: apm.transaction
+dataset_is_prefix: true
+ilm_policy: metrics-apm.internal_metrics_transaction-default_policy
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics_transaction/manifest.yml:5]: Should we change the directory name to "transaction_metrics", to be more in line with the data stream name? Then the ILM policy name would be more consistent too.

File: apmpackage/apm/data_stream/internal_metrics_transaction/manifest.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+elasticsearch:
+  index_template:
+    mappings:
+      # Internal metrics should have all fields strictly mapped;
+      # we are in full control of the field names.
```
Comment (axw) [inline apmpackage/apm/data_stream/internal_metrics_transaction/manifest.yml:10]:
```suggestion
      # Transaction metrics should have all fields strictly mapped;
      # we are in full control of the field names.
```

**Review by axw — COMMENTED** [review]
LGTM from a high level. Just a few minor comments. I haven't pored over the txmetrics/aggregator.go code changes.

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go (hunk 2 above)
Comment (marclop) [reply x-pack/apm-server/aggregation/txmetrics/aggregator.go:58]: I looked into making the change but in my opinion it made the code slightly less readable and cryptic. I'm happy to revisit if you feel it would be better to move it to a slice.

Comment (axw) [reply x-pack/apm-server/aggregation/txmetrics/aggregator.go:58]: Totally fine, thanks for looking into it.

**Review by felixbarny — APPROVED** [review]
LGTM. Note to self and others: `event.duration` is mapped as a constant_keyword and describes the aggregation interval.
One question I have is how the UI is supposed to know which intervals to target as the aggregation intervals are configurable. Should it do a terms_enum on the `event.duration`, query for `metrics-apm.transaction-*` data streams, or hard-code the intervals?

**Comment by marclop** [comment]
@felixbarny I was wondering about that myself too. I had a couple of thoughts on what to do, my understanding is that the UI will attempt to use `10m` buckets on longer time ranges, I don't know how they'll _discover_ which time ranges are available, and how to fill them, Aggregation intervals are available as:

- Templated on the index patterns (`metrics-apm.transaction.{1m,10m,60m}`)
- In each individual document as `event.duration` as `constant_keyword` instead of `long` type.

**Comment by felixbarny** [comment]
> Templated on the index patterns (`metrics-apm.transaction.{1m,10m,60m}`)

Can the UI rely on these granularities or can users modify that, so that there are these data streams `metrics-apm.transaction.{30s,5m,30m}`, for example?

**Comment by marclop** [comment]
@felixbarny we discussed this in the stand-up and have an agreement that we should disallow users to change any of the aggregation intervals going forward. This will be a breaking change but will guarantee the UI behavior going forward.

Another open point is that the current data stream definitions don't allow dedicated ILM policies, I'll investigate how we can define an ILM policy per interval without making the APM package definitions look too bloated.

**Comment by felixbarny** [comment]
> I'll investigate how we can define an ILM policy per interval without making the APM package definitions look too bloated.

To avoid duplicating the field definitions, can we use symlinks in package definitions?

**Comment by marclop** [comment]
I've updated the code to:

- Create 3 different data streams from 3 intervals using templates. This allows uses to configure the retention policy for the different periods. By default I've configured some rollover and delete settings in the different intervals, please let me know if we should change the retention windows. 
- Disallow configuring the aggregation intervals for transaction metrics.

**Comment by felixbarny** [comment]
> Create 3 different data streams from 3 intervals using templates. 

I see that the ILM policies are so that we keep 30d@1m, 90d@10m, and 180d@60m. Previously, we were storing 90d@1m. What's the relative overall storage overhead or storage saving that results from this change?

**Comment by marclop** [comment]
@felixbarny, I haven't run the numbers. I had assumed higher resolution for most cases for 30d, then lower resolution for 90d, and even more up to 180. I'm happy to keep it 90d, 180d and 360d if you feel that'd be more appropriate.

**Comment by dgieselaar** [comment]
@marclop we need a constant keyword in the index that denotes the aggregation interval to select the right data stream. I've used `metricset.interval` in my tests but I've just made that up, maybe there's an ECS-compatible alternative?

**Comment by marclop** [comment]
@dgieselaar. I've "re-purposed" `event.duration` for all the `metrics-apm.transaction.{1m,10m,60m}` data streams. The raw value is stored as a `constant_keyword` in nanoseconds.

https://github.com/elastic/apm-server/blob/242f4d8bf6c80eb362b9ad9c69b8e96938afc075/systemtest/approvals/TestTransactionAggregation.approved.json#L19

**Comment by dgieselaar** [comment]
@marclop it would be easier for us if it was a term, like `1m` or `10m`. Also for custom dashboards, e.g. consider the difference between `event.duration: 3600000000000` and `metricset.interval:60m`. I realise `event.duration` should be nanoseconds but maybe we can use something else?

**Comment by felixbarny** [comment]
IINM, we should be saving around 53% with these settings:

```
30/1m + 90/10m + 180/60m=42 day's worth of 1m buckets
1-(42@1m/90@1m)=53.3%
```

**Comment by felixbarny** [comment]
If we wanted, we could even add an additional layer of 10s granularity stored for 7d and still use less storage than before.

**Comment by marclop** [comment]
@dgieselaar Regarding the format of this new `metricset.interval` field, should it always be formatted in minutes with the `m` unit? Or are we planning to use different units based on the value?

I'm asking because I think we should define which unit is the minimum and maximum unit for that field. I've created a small snipped so you can run it and see different outputs: https://go.dev/play/p/V31MJ_SQNHr.

`60 minutes` could be formatted either `1h0m0s`, `1h` or `60m`. If we introduce intervals lower than 1m, then the problem expands.

I think we should store either the value in seconds (`50s`, `10s`, etc) or in minutes if `value >= minute` (`1m`, `60m`, `720m` (`12h`)).

**Comment by marclop** [comment]
I've opened https://github.com/elastic/apm-data/pull/7 to update the model to include the new field.

**Comment by dgieselaar** [comment]
@marclop re:
> I think we should store either the value in seconds (50s, 10s, etc) or in minutes if value >= minute (1m, 60m, 720m (12h)).

This sounds good to me. I'm also fine with 1h, 12h etc beyond a 1h interval.

**Comment by simitt** [comment]
>Looking good! As discussed on Slack, we'll need to handle transaction metrics from older servers going to metrics-apm.internal-*, and either upgrade and re-route, or drop them.

Apologies for being late to this discussion, but dropping them would basically mean that as soon as one apm-server is upgraded, all transaction metrics from not (yet) upgraded APM Servers would be dropped. The APM UI needs to be able to handle the case that tx metrics are available in different data streams anyways, why do we drop transaction metrics that are going to the _old_ datastreams, rather than making it a UI problem how to deal with them? As soon as all APM Servers are upgraded, the _old_ tx metrics won't be there anyways, up until then, why not just ignoring them but preserving the data? Dropping seems radical, especially in a minor.

File: apmpackage/apm/data_stream/transaction_interval_metrics/elasticsearch/ilm/default_policy.60m.json [hunk 10]
```diff
@@ -0,0 +1,23 @@
+{
+    "policy": {
+        "phases": {
+            "hot": {
+                "actions": {
+                    "rollover": {
+                        "max_age": "30d",
+                        "max_size": "20gb"
```
[the remark this answers is not in the document]
Comment (marclop) [reply apmpackage/apm/data_stream/transaction_interval_metrics/elasticsearch/ilm/default_policy.60m.json:8]: I've changed the `max_size` to `50gb` as before. I can't remember why I changed the max size, but I do remember thinking that 50gb worth of metrics (especially only `transaction metrics`), seemed like quite a large index.

I did lower the `max_age` for the smaller intervals:

- `1m` => `7d`
- `10m` => `14d`
- `60m` => `30d` (default)

**Comment by marclop** [comment]
> Apologies for being late to this discussion, but dropping them would basically mean that as soon as one apm-server is upgraded, all transaction metrics from not (yet) upgraded APM Servers would be dropped. The APM UI needs to be able to handle the case that tx metrics are available in different data streams anyways, why do we drop transaction metrics that are going to the old datastreams, rather than making it a UI problem how to deal with them? As soon as all APM Servers are upgraded, the old tx metrics won't be there anyways, up until then, why not just ignoring them but preserving the data? Dropping seems radical, especially in a minor.

@simitt Since we're removing the transaction, service and service destination metrics from the `metrics-apm.internal`, my understanding is that we'll remove the fields that are specific to those metrics (tx, service and service destination) from `metrics-apm.internal` as well, because of that, each PR that moves those metrics to its own dedicated interval data stream, should include an ingest pipeline that drops those metric sets.

We reasoned with @axw that only the last `1m` interval metrics would be lost (time between the new integration package being installed and the APM Servers being upgraded) and as such, it's a very small amount of data "lost".

**Comment by simitt** [comment]
> We reasoned with @axw that only the last 1m interval metrics would be lost (time between the new integration package being installed and the APM Servers being upgraded) and as such, it's a very small amount of data "lost".

This makes sense for the upgrade scenario, which will cover a majority of usecases. I wonder if the setup of multiple APM Servers of different versions writing to the same Elasticsearch instance is a real scenario or a theoretical problem. 
How much work would it be to add a pipeline that sends the 1m interval metrics from older APM Servers to the new data stream? The _old_ and the _new_ 1m interval datastreams would be compatible afaics.

**Comment by marclop** [comment]
We discussed the drop pipeline with @simitt and came to the conclusion that the best course of action is leaving the `metrics-apm.internal` as is without a drop pipeline, let users upgrade and introduce a drop pipeline in an upcoming version providing a clear upgrade path so metrics aren't lost.

**Comment by marclop** [comment]
@simitt I believe I've updated all the docs for metrics. Let me know if I've missed anything

**Review by simitt — APPROVED** [review]
Changes LGTM! 

Please also update the [dev-docs on transaction metrics](https://github.com/elastic/apm-server/blob/main/dev_docs/trace_metrics.md#transaction-metrics). You can do that in a follow up, when working on the other aggregation metrics.

**Comment by marclop** [comment]
systemtest is flaky, merging anyway
